File: ThrowExtensions.cs
using System;
using System.Runtime.CompilerServices;

#pragma warning disable IDE0060 // Remove unused parameter

namespace Platform.Exceptions
{
    public static class ThrowExtensions
    {
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static void NotSupportedException(this ThrowExtensionRoot root) => throw new NotSupportedException();

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static T NotSupportedExceptionAndReturn<T>(this ThrowExtensionRoot root) => throw new NotSupportedException();

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static void NotImplementedException(this ThrowExtensionRoot root) => throw new NotImplementedException();

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static T NotImplementedExceptionAndReturn<T>(this ThrowExtensionRoot root) => throw new NotImplementedException();
    }
}

File: cpp/Platform.Exceptions/ThrowExtensions.cpp
namespace Platform::Exceptions
{
    class ThrowExtensions
    {
        public: static void NotSupportedException(Platform::Exceptions::ExtensionRoots::ThrowExtensionRoot root) { throw std::logic_error("Not supported exception."); }

        public: template <typename TReturn> static TReturn NotSupportedExceptionAndReturn(Platform::Exceptions::ExtensionRoots::ThrowExtensionRoot root) { throw std::logic_error("Not supported exception."); }

        public: static void NotImplementedException(Platform::Exceptions::ExtensionRoots::ThrowExtensionRoot root) { throw std::logic_error("Not implemented exception."); }

        public: template <typename TReturn> static TReturn NotImplementedExceptionAndReturn(Platform::Exceptions::ExtensionRoots::ThrowExtensionRoot root) { throw std::logic_error("Not implemented exception."); }
    };
}
